test/fslock.cpp

```diff
@@ -0,0 +1,152 @@
+#include "fslock.h"
+
+#include <fcntl.h>
+#include <semaphore.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "3rd-party/catch.hpp"
+#include "test-helpers.h"
+
+using namespace newsboat;
+
+// Forks and calls FsLock::try_lock() in the child process.
+class LockProcess {
```

Comment:
Can we use a thread for this, instead of a process? A thread runs a single function, which neatly avoids the problem with destructors.

Reply:
I don't think that is possible.

The [`FsLock` implementation](https://github.com/newsboat/newsboat/blob/56b336b74e1a91ea89e6a60755dc3b0a57fbc8e3/src/fslock.cpp#L45) uses `lockf`.
Those locks are registered per process so locking in a different thread will not block the current thread from locking the same file.

Reply:
Oh. Starting the process from a thread, then? The process will copy the stack of the thread, which will contain just one frame and a single local (and `int` to store the result of `fork()`).

I'm just worried that running a second process with a full copy of tests' memory, and then not running a destructor of FsLock in there, might create some unobvious problems that will either make tests brittle, or make them behave differently from "real world". Simply because this setup is notably different from the "real world" where two separate processes try to lock the same file.

I can't come up with a specific problem, though, so this isn't a blocker. I can merge this PR as-is if you want, and work on the process-from-thread idea myself. Should I?

Reply:
Sounds like a good idea.

> I can merge this PR as-is if you want, and work on the process-from-thread idea myself. Should I?

If you want to, I would be happy with that.
Otherwise I might do that somewhere later this week.

Reply:
Okay! PR is merged, and the process-from-thread idea is moved to #790.

Reply:
Nice to see such a clear explanation in that issue!